File: src/main/scala/fi/oph/koski/tutkinto/TutkintoRakenne.scala
package fi.oph.koski.tutkinto

import fi.oph.koski.schema.{Koodistokoodiviite, LocalizedString}
import fi.oph.koski.tutkinto.Koulutustyyppi.Koulutustyyppi

case class TutkintoRakenne(
  id: Long,
  diaarinumero: String,
  koulutustyyppi: Koulutustyyppi,
  suoritustavat: List[SuoritustapaJaRakenne],
  osaamisalat: List[Koodistokoodiviite],
  tutkintonimikkeet: List[Koodistokoodiviite],
  koulutukset: List[Koodistokoodiviite]
) {
  def findSuoritustapaJaRakenne(suoritustapa: Koodistokoodiviite): Option[SuoritustapaJaRakenne] = {
    suoritustavat.find(_.suoritustapa == suoritustapa)
  }
}

case class SuoritustapaJaRakenne(suoritustapa: Koodistokoodiviite, rakenne: Option[RakenneOsa])
case class TutkinnonOsanLaajuus(min: Option[Long], max: Option[Long])

sealed trait RakenneOsa {
  def sisältääMäärittelemättömiäOsia: Boolean
  def tutkinnonOsat: List[TutkinnonOsa]
}

case class RakenneModuuli(nimi: LocalizedString, osat: List[RakenneOsa], määrittelemätön: Boolean, laajuus: Option[TutkinnonOsanLaajuus]) extends RakenneOsa {
  def tutkinnonOsat: List[TutkinnonOsa] = osat flatMap {
    case m: RakenneModuuli => m.tutkinnonOsat
    case o: TutkinnonOsa => List(o)
  }
  def tutkinnonRakenneLaajuus: TutkinnonOsanLaajuus = {
    this.laajuus.getOrElse(TutkinnonOsanLaajuus(None, None))
  }

  override def sisältääMäärittelemättömiäOsia: Boolean = määrittelemätön || osat.exists(_.sisältääMäärittelemättömiäOsia)
}
case class TutkinnonOsa(tunniste: Koodistokoodiviite, nimi: LocalizedString, laajuus: Option[Long], osaAlueet: List[TutkinnonOsanOsaAlue]) extends RakenneOsa {
  override def tutkinnonOsat: List[TutkinnonOsa] = List(this)
  override def sisältääMäärittelemättömiäOsia: Boolean = false
  // TODO add laajuus (from viite) + osa-alueet (+ thier laajuus + pakollisuus)
}

case class TutkinnonOsanOsaAlue(id: Long, koodiarvo: String, pakollisenOsanLaajuus: Option[Long], valinnaisenOsanLaajuus: Option[Long])

File: web/app/types/fi/oph/koski/tutkinto/TutkintoRakenne.ts
import { Koodistokoodiviite } from '../schema/Koodistokoodiviite'
import { SuoritustapaJaRakenne } from './SuoritustapaJaRakenne'

/**
 * TutkintoRakenne
 *
 * @see `fi.oph.koski.tutkinto.TutkintoRakenne`
 */
export type TutkintoRakenne = {
  $class: 'fi.oph.koski.tutkinto.TutkintoRakenne'
  tutkintonimikkeet: Array<Koodistokoodiviite>
  diaarinumero: string
  id: number
  suoritustavat: Array<SuoritustapaJaRakenne>
  osaamisalat: Array<Koodistokoodiviite>
  koulutustyyppi: Koodistokoodiviite
  koulutukset: Array<Koodistokoodiviite>
  voimassaoloAlkaa?: string
}

export const TutkintoRakenne = (o: {
  tutkintonimikkeet?: Array<Koodistokoodiviite>
  diaarinumero: string
  id: number
  suoritustavat?: Array<SuoritustapaJaRakenne>
  osaamisalat?: Array<Koodistokoodiviite>
  koulutustyyppi: Koodistokoodiviite
  koulutukset?: Array<Koodistokoodiviite>
  voimassaoloAlkaa?: string
}): TutkintoRakenne => ({
  tutkintonimikkeet: [],
  suoritustavat: [],
  osaamisalat: [],
  $class: 'fi.oph.koski.tutkinto.TutkintoRakenne',
  koulutukset: [],
  ...o
})

TutkintoRakenne.className = 'fi.oph.koski.tutkinto.TutkintoRakenne' as const

export const isTutkintoRakenne = (a: any): a is TutkintoRakenne =>
  a?.$class === 'fi.oph.koski.tutkinto.TutkintoRakenne'
